pupil_src/shared_modules/gaze_producer/ui/timeline.py

```diff
@@ -0,0 +1,80 @@
+"""
+(*)~---------------------------------------------------------------------------
+Pupil - eye tracking platform
+Copyright (C) 2012-2018 Pupil Labs
+
+Distributed under the terms of the GNU
+Lesser General Public License (LGPL v3.0).
+See COPYING and COPYING.LESSER for license details.
+---------------------------------------------------------------------------~(*)
+"""
+from plugin_timeline import Line, BarsElementTs, RangeElementPerc
+
+
+class Timeline:
+    def __init__(
+        self,
+        plugin_timeline,
+        reference_detection_controller,
+        reference_location_storage,
+        plugin,
+    ):
+        self._plugin_timeline = plugin_timeline
+        self._reference_detection_controller = reference_detection_controller
+        self._reference_location_storage = reference_location_storage
+
+        plugin.add_observer("init_ui", self._on_init_ui)
+        self._reference_detection_controller.add_observer(
+            "start_detection", self._on_start_reference_detection
+        )
+        self._reference_location_storage.add_observer(
+            "add", self._on_reference_storage_changed
+        )
+        self._reference_location_storage.add_observer(
+            "delete", self._on_reference_storage_changed
+        )
+        self._reference_location_storage.add_observer(
+            "delete_all", self._on_reference_storage_changed
+        )
```

Comment:
That's an interesting point. I would actually want to make this a style guide, I mean almost always using event handlers instead of direct calls.

Goal: Make it easy to see all events a component is reacting to and what it does in this case.

If the actions would always be as trivial as calling a single function, then this `add_observer` block in `__init__` would exactly provide this. It's slightly ugly, but you could find all info there. However, in many cases, the actions are not that simple, so you need event handlers in some cases. So I always use event handlers, so you can look at all `_on_something` methods at the end of a class to see all events its reacting to and what happens.

Bonuses:
- you can observe event handlers (e.g. event -> controller changes X because of the event -> ui refreshes because X changed)
- these method definitions are easier to read, compare:
```
self._reference_location_storage.add_observer(
    "add", self.render()
)
self._reference_location_storage.add_observer(
    "delete", self.render()
)
self._reference_location_storage.add_observer(
    "delete_all", self.render()
)
```
vs
```
def _on_reference_storage_changed(self, *args, **kwargs):
    self.render()
```